flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/PulsarSourceReaderFactory.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.flink.connector.pulsar.source.reader.source.PulsarUnorderedSourceReader;
+import org.apache.flink.connector.pulsar.source.reader.split.PulsarOrderedPartitionSplitReader;
+import org.apache.flink.connector.pulsar.source.reader.split.PulsarUnorderedPartitionSplitReader;
+import org.apache.flink.connector.pulsar.source.split.PulsarPartitionSplit;
+
+import org.apache.flink.shaded.guava30.com.google.common.io.Closer;
+
+import org.apache.pulsar.client.admin.PulsarAdmin;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.SubscriptionType;
+import org.apache.pulsar.client.api.transaction.TransactionCoordinatorClient;
+import org.apache.pulsar.client.impl.PulsarClientImpl;
+import org.apache.pulsar.client.impl.conf.ConsumerConfigurationData;
+
+import java.util.function.Supplier;
+
+import static org.apache.flink.connector.base.source.reader.SourceReaderOptions.ELEMENT_QUEUE_CAPACITY;
+import static org.apache.flink.connector.pulsar.common.utils.PulsarClosingUtils.closeWithTimeout;
+
+/**
+ * This factory class is used for creating different types of source reader for different
+ * subscription type.
+ *
+ * <ol>
+ *   <li>Failover, Exclusive: We would create PulsarOrderedSourceReader.
+ *   <li>Shared, Key_Shared: We would create PulsarUnorderedSourceReader.
+ * </ol>
+ */
+public final class PulsarSourceReaderFactory {
+
+    private PulsarSourceReaderFactory() {
+        // No public constructor.
+    }
+
+    public static <IN, OUT> SourceReader<OUT, PulsarPartitionSplit> create(
+            SourceReaderContext readerContext,
+            PulsarDeserializationSchema<IN, OUT> deserializationSchema,
+            PulsarClient pulsarClient,
+            PulsarAdmin pulsarAdmin,
+            Configuration configuration,
+            SourceConfiguration sourceConfiguration,
+            ConfigurationDataCustomizer<ConsumerConfigurationData<IN>>
+                    consumerConfigurationCustomizer) {
+        // Create a message queue with the predefined source option.
+        FutureCompletingBlockingQueue<RecordsWithSplitIds<PulsarMessage<OUT>>> elementsQueue =
+                new FutureCompletingBlockingQueue<>(
+                        configuration.getInteger(ELEMENT_QUEUE_CAPACITY));
+
+        // Create the required close component for pulsar source reader.
+        Closer splitCloser = Closer.create();
+
+        splitCloser.register(
+                () ->
+                        closeWithTimeout(
+                                "PulsarClient",
+                                pulsarClient::close,
+                                sourceConfiguration.getCloseTimeoutMs()));
+        splitCloser.register(
+                () ->
+                        closeWithTimeout(
+                                "PulsarAdmin",
+                                pulsarAdmin::close,
+                                sourceConfiguration.getCloseTimeoutMs()));
```

Comment:
Why do we need to add this here? Can't we just close them in the reader (it's owned by the reader after all).  Since Closer is Closeable, you could have a Closer in the reader if you want to preserve order and get the free suppression chaining.